persist/fs/commitlog/reader.go

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package commitlog
+
+import (
+	"bufio"
+	"encoding/binary"
+	"errors"
+	"hash/adler32"
+	"io"
+	"os"
+	"time"
+
+	"github.com/m3db/m3db/generated/proto/schema"
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3x/time"
+
+	"github.com/golang/protobuf/proto"
+)
+
+var (
+	errCommitLogReaderAlreadyOpen               = errors.New("commit log reader already open")
+	errCommitLogReaderChunkSizeChecksumMismatch = errors.New("commit log reader encountered chunk size checksum mismatch")
+	errCommitLogReaderChunkDataChecksumMismatch = errors.New("commit log reader encountered chunk data checksum mismatch")
+	errCommitLogReaderMissingLogMetadata        = errors.New("commit log reader encountered message missing metadata")
+)
+
+type commitLogReader interface {
+	io.Closer
+
+	// Open opens the commit log for reading
+	Open(filePath string) (time.Time, time.Duration, int, error)
+
+	// Read returns the next key and data pair or error, will return io.EOF at end of volume
+	Read() (m3db.CommitLogSeries, m3db.Datapoint, xtime.Unit, []byte, error)
+}
+
+type reader struct {
+	opts           m3db.DatabaseOptions
+	chunkReader    chunkReader
+	chunkReaderRef *chunkReader
+	sizeBuffer     []byte
+	dataBuffer     []byte
+	info           schema.CommitLogInfo
+	log            schema.CommitLog
+	metadata       schema.CommitLogMetadata
+	metadataLookup map[uint64]m3db.CommitLogSeries
+}
+
+func newCommitLogReader(opts m3db.DatabaseOptions) commitLogReader {
+	r := &reader{
+		opts: opts,
+		chunkReader: chunkReader{
+			buffer: bufio.NewReaderSize(nil, bufferWriteSize),
+		},
+		sizeBuffer:     make([]byte, binary.MaxVarintLen64),
+		metadataLookup: make(map[uint64]m3db.CommitLogSeries),
+	}
+	r.chunkReaderRef = &r.chunkReader
```

Comment:
perhaps we can have a `newChunkReader` method that returns a pointer so we don't need both the value and the pointer?